lib/baza-rb.rb

```diff
@@ -579,6 +580,38 @@ def csrf
 
   private
 
+  # Update host from X-Zerocracy-Host header if present.
+  #
+  # @param [Typhoeus::Response] ret The HTTP response containing headers
+  # @param [Iri] uri The current URI object to update
+  # @return [Iri] The updated URI object (or original if no valid header present)
+  # @note Invalid hostnames are logged as warnings and ignored
+  def update_host_from_response(ret, uri)
+    sticky = ret.headers && ret.headers['X-Zerocracy-Host']
+    return uri unless sticky
+    return uri unless valid_hostname?(sticky)
+    new_host = sticky.strip.chomp('.')
+    @host_mutex.synchronize do
+      if new_host != @host
+        @loog.debug("Switching host from #{@host} to #{new_host} as per X-Zerocracy-Host")
+        @host = new_host
+      end
+      uri = uri.host(@host)
+    end
+    uri
+  end
+
+  # Validate hostname format according to RFC 1123.
+  #
+  # @param [String] hostname The hostname to validate
+  # @return [Boolean] True if valid, false otherwise
+  def valid_hostname?(name)
+    return false unless name.is_a?(String)
```

Comment:
@paveltrty how "non string" may happen in this private method? see https://github.com/yegor256/at?tab=readme-ov-file#3-31